File: examples/python/opencap_import.py
"""Model Health Python SDK — Import an OpenCap session into Model Health.

Usage:
    opencap_import.py <api_key> <opencap_token> <opencap_session_id>
"""

import json

from docopt import docopt

import time

from modelhealth import (
    ModelHealthService,
    SubjectParameters,
    SessionConfig,
    SessionOpenSimModel,
    SessionScalingSetup,
    SessionCoreEngine,
    FilterFrequencyDefault,
    FilterFrequencyHz,
    ImportStatusCreatedSession,
    ImportStatusUploading,
    ActivityType,
    ActivityStatus,
    ActivityStatusUploading,
    ActivityStatusAnalyzing,
    AnalysisStatus,
)
from _opencap_api import fetch_session, fetch_subject
from _prompts import confirm, pick_one


_OPENSIM_MODEL_MAP = {
    "LaiUhlrich2022_shoulder": SessionOpenSimModel.lai_uhlrich_2022_shoulder,
    "LaiUhlrich2022": SessionOpenSimModel.lai_uhlrich_2022,
}

_SCALING_SETUP_MAP = {
    "upright_standing_pose": SessionScalingSetup.upright_standing_pose,
    "any_pose": SessionScalingSetup.any_pose,
}

_CORE_ENGINE_MAP = {
    "v0.2": SessionCoreEngine.v0_2,
    "v0.3": SessionCoreEngine.v0_3,
    "v1.0": SessionCoreEngine.v1_0,
}


def _filter_frequency_from_meta(value):
    if value is None or value == "default":
        return FilterFrequencyDefault()
    try:
        return FilterFrequencyHz(int(value))
    except (TypeError, ValueError):
        return FilterFrequencyDefault()


def _pick_trial(trials, name_filter):
    """Return the best matching trial: prefer done status, then latest created_at."""
    matches = [t for t in trials if name_filter(t["name"].lower())]
    done = [t for t in matches if t.get("status") == "done"]
    pool = done if done else matches
    if not pool:
        return None
    return max(pool, key=lambda t: t.get("created_at", ""))



def copy_session(
    mh_service,
    session_id_input,
    user_token_input,
    api_url_input="https://api.opencap.ai/",
    meta_overrides=None,
    trials_to_import=None,
    trial_meta_overrides=None,
):
    """Copy an OpenCap session into a ModelHealth account.

    Downloads all trials from the given OpenCap session and re-uploads them to
    a new ModelHealth session via :meth:`ModelHealthService.import_session`.

    Args:
        mh_service: An active :class:`ModelHealthService` instance.
        session_id_input: UUID of the OpenCap session to copy from.
        user_token_input: OpenCap authentication token.
        api_url_input: Base URL of the OpenCap API.
        meta_overrides: Optional dict of settings to set under meta["settings"] on the new
            session, overriding or extending what is copied from the source session. Supported keys:
                - "openSimModel": "LaiUhlrich2022_shoulder" | "LaiUhlrich2022"
                - "scalingsetup": "upright_standing_pose" | "any_pose"
                - "coreengine": "v0.2" | "v0.3" | "v1.0"
                - "filterfrequency": "default" | <integer>
        trials_to_import: Optional list of trial names to import. Setup trials
            (calibration, neutral) are always included. Set to None to import all trials.
        trial_meta_overrides: Optional dict keyed by trial name. Each value is a dict of
            per-trial overrides. Supported keys:
                - "activity_type": ActivityType — triggers automatic analysis after processing.
    """
    # Fetch source session
    session_input = fetch_session(session_id_input, api_url=api_url_input, api_token=user_token_input)

    # Step 1: Build session config from source settings, applying any overrides
    meta = session_input.get("meta") or {}
    if isinstance(meta, str):
        meta = json.loads(meta) if meta else {}
    settings = dict(meta.get("settings") or {})
    if meta_overrides:
        settings.update(meta_overrides)
        print(f"Applying settings overrides: {meta_overrides}")

    session_config = SessionConfig(
        opensim_model=_OPENSIM_MODEL_MAP.get(settings.get("openSimModel"), SessionOpenSimModel.lai_uhlrich_2022_shoulder),
        scaling_setup=_SCALING_SETUP_MAP.get(settings.get("scalingsetup"), SessionScalingSetup.upright_standing_pose),
        core_engine=_CORE_ENGINE_MAP.get(settings.get("coreengine"), SessionCoreEngine.v0_3),
        filter_frequency=_filter_frequency_from_meta(settings.get("filterfrequency")),
    )
    print("Session settings configured.")

    # Step 2: Select or create subject
    subjects = mh_service.subject_list()
    if subjects and confirm(f"Found {len(subjects)} subject(s). Select an existing one?", default=True):
        print()
        subject = pick_one(subjects, "Select subject", lambda s: f"{s.name}  (ID {s.id})")
        print(f"  Using: {subject.name}")
    else:
        # Copy subject from the OpenCap session.
        # OpenCap stores height in metres; SubjectParameters expects centimetres.
        subject_data = fetch_subject(session_input["subject"], api_url=api_url_input, api_token=user_token_input)
        params = SubjectParameters(
            name=subject_data.get("name") or subject_data.get("first_name", ""),
            weight=float(subject_data.get("weight") or 0),
            height=float(subject_data.get("height") or 0) * 100,
            birth_year=subject_data.get("birth_year"),
        )
        subject = mh_service.create_subject(params)
        print("Subject created successfully.")

    # Step 3: Resolve calibration — prefer parent session referenced in meta, fall back to current session
    calibration_trial = None
    parent_calibration_id = (meta.get("sessionWithCalibration") or {}).get("id")

    if parent_calibration_id:
        try:
            parent_session = fetch_session(parent_calibration_id, api_url=api_url_input, api_token=user_token_input)
            cal = _pick_trial(parent_session["trials"], lambda n: n == "calibration")
            if cal is not None:
                calibration_trial = cal
                print(f"Using calibration from parent session {parent_calibration_id}.")
        except Exception as e:
            print(f"Could not load calibration from parent session: {e}")

    if calibration_trial is None:
        cal = _pick_trial(session_input["trials"], lambda n: n == "calibration")
        if cal is not None:
            calibration_trial = cal

    # Build ordered trial list: calibration -> neutral -> dynamic
    # Skip setup trials that aren't done on the source side.
    neutral_trial = _pick_trial(session_input["trials"], lambda n: "neutral" in n)
    dynamic_trials = [
        t for t in session_input["trials"]
        if t["name"].lower() not in ("calibration", "neutral")
        and (trials_to_import is None or t["name"] in trials_to_import)
    ]
    if calibration_trial is None:
        print("No calibration trial found, cannot proceed.")
        return
    if calibration_trial.get("status") != "done":
        print("Calibration trial is not done, cannot proceed.")
        return
    if neutral_trial is None:
        print("No neutral trial found, cannot proceed.")
        return
    if neutral_trial.get("status") != "done":
        print("Neutral trial is not done, cannot proceed.")
        return

    # Step 4: Transfer all trials in one call
    # Build the full source session meta with settings overrides applied so the
    # backend has all the context it needs (e.g. sessionWithCalibration).
    session_meta = dict(meta)
    session_meta.setdefault("settings", {}).update(meta_overrides or {})

    # Annotate trials with per-trial overrides (e.g. activity_type).
    # Setup trials are never annotated — only dynamic trials support activity_type.
    def _annotate_trial(trial):
        print(f"Preparing trial {trial['name']} for import...")
        overrides = (trial_meta_overrides or {}).get(trial["name"], {})
        if not overrides:
            print(f"No overrides for trial {trial['name']}, importing as-is.")
            return trial
        trial = dict(trial)
        if "activity_type" in overrides:
            print(f"Applying activity_type {overrides['activity_type']} to trial {trial['name']}")
            trial["activity_type"] = str(overrides["activity_type"])
        else:
            print(f"No activity_type override for trial {trial['name']}, importing as-is.")
        return trial

    annotated_trials = [_annotate_trial(t) for t in dynamic_trials]

    _seen_uploads = {}
    _current_trial = [None]
    def on_status(status):
        if status == "creating_session":
            print("  Creating session...")
        elif isinstance(status, ImportStatusCreatedSession):
            print(f"  Session created: {status.session_id}")
        elif isinstance(status, ImportStatusUploading) and status.uploaded < status.total:
            _current_trial[0] = status.trial
            if _seen_uploads.get(status.trial) != status.uploaded:
                _seen_uploads[status.trial] = status.uploaded
                print(f"  [{status.trial}] Uploading video {status.uploaded + 1}/{status.total}...")
        elif status == "processing":
            trial = _current_trial[0] or "activity"
            print(f"  [{trial}] Processing...")

    session = mh_service.import_session(
        json.dumps([calibration_trial, neutral_trial, *annotated_trials]),
        subject,
        session_config=session_config,
        session_meta_json=json.dumps(session_meta),
        status_callback=on_status,
    )
    print("All activities imported successfully to session: ", session.id)
    return session


def wait_for_activities(mh_service, session, poll_interval=5):
    """Poll activity status for all dynamic activities until each is ready or failed.

    Args:
        mh_service: An active :class:`ModelHealthService` instance.
        session: The Session returned by :func:`copy_session`.
        poll_interval: Seconds between status polls. Default: 5.
    """
    pending = [a for a in session.activities if a.name.lower() not in ("calibration", "neutral")]

    if not pending:
        return

    # Phase 1: wait for processing to finish.
    # Activities that move to Analyzing are collected for a second pass.
    print("Waiting for activities to process...")
    analyzing = []  # list of (activity, task) pairs
    while pending:
        still_pending = []
        for activity in pending:
            status = mh_service.activity_status(activity)
            if isinstance(status, ActivityStatusUploading):
                print(f"  [{activity.name}] Uploading ({status.uploaded}/{status.total})...")
                still_pending.append(activity)
            elif status == ActivityStatus.processing:
                print(f"  [{activity.name}] Processing...")
                still_pending.append(activity)
            elif isinstance(status, ActivityStatusAnalyzing):
                print(f"  [{activity.name}] Analyzing — queued for analysis poll.")
                analyzing.append((activity, status.task))
            elif status == ActivityStatus.ready:
                print(f"  [{activity.name}] Ready.")
            elif status == ActivityStatus.failed:
                print(f"  [{activity.name}] Failed.")
        if still_pending:
            time.sleep(poll_interval)
        pending = still_pending

    # Phase 2: poll analysis status for activities that are analyzing.
    if not analyzing:
        return

    print("Waiting for analyses to complete...")
    pending_analysis = list(analyzing)
    while pending_analysis:
        still_pending = []
        for activity, task in pending_analysis:
            status = mh_service.analysis_status(task)
            if status == AnalysisStatus.processing:
                print(f"  [{activity.name}] Analyzing...")
                still_pending.append((activity, task))
            elif status == AnalysisStatus.completed:
                print(f"  [{activity.name}] Analysis complete.")
            else:
                print(f"  [{activity.name}] Analysis failed.")
        if still_pending:
            time.sleep(poll_interval)
        pending_analysis = still_pending


if __name__ == "__main__":
    args = docopt(__doc__)

    # Optional: override or extend session metadata with ModelHealth-specific settings.
    # Leave as None to keep the values from the source session.
    # We recommend setting the "openSimModel" to "LaiUhlrich2022_shoulder" and "coreengine" to "v1.0" for best results.
    # The "filterfrequency" setting can be set to an integer value in Hz to apply a low-pass Butterworth filter at that frequency.
    meta_overrides = {
        "openSimModel": "LaiUhlrich2022_shoulder",  # "LaiUhlrich2022_shoulder" | "LaiUhlrich2022"
        "scalingsetup": "upright_standing_pose",    # "upright_standing_pose" | "any_pose"
        "coreengine": "v1.0",                       # "v0.2" | "v0.3" | "v1.0"
        "filterfrequency": "default",               # "default" | <integer Hz>
    }

    # Trials to import. Setup trials (calibration, neutral) are always included.
    # Set to None to import all trials.
    trials_to_import = ["test"]

    # Optional per-trial overrides keyed by trial name.
    # Setting "activity_type" will auto-launch analysis after processing.
    trial_meta_overrides = {
        "test":     {"activity_type": ActivityType.range_of_motion},
    }

    mh_service = ModelHealthService(args["<api_key>"])
    session = copy_session(
        mh_service,
        session_id_input=args["<opencap_session_id>"],
        user_token_input=args["<opencap_token>"],
        meta_overrides=meta_overrides,
        trials_to_import=trials_to_import,
        trial_meta_overrides=trial_meta_overrides,
    )
    wait_for_activities(mh_service, session)

File: examples/ts/opencap_import.ts
/**
 * Model Health TypeScript SDK — import an OpenCap session into Model Health.
 * Mirrors examples/python/opencap_import.py.
 *
 * Usage:
 *   npx tsx opencap_import.ts [<api_key>] [<opencap_session_id>] [<opencap_token>]
 *
 * Arguments are resolved from CLI args, .env file, or environment variables:
 *   MODEL_HEALTH_API_KEY  — Model Health API key
 *   OPENCAP_SESSION_ID   — OpenCap session UUID
 *   OPENCAP_TOKEN        — OpenCap authentication token (or prompted via login)
 *
 * Edit metaOverrides and trialMetaOverrides below to customise the import.
 */

import { ModelHealthService, ActivityType } from '@modelhealth/modelhealth';
import type {
  Subject, Session, Activity, Analysis,
  SessionConfig, SessionOpenSimModel, SessionScalingSetup, SessionCoreEngine,
  FilterFrequency, ImportStatus, ActivityStatus,
} from '@modelhealth/modelhealth';
import {
  loadApiKey, dotEnvValue, prompt, pickOne, confirm, sleep, closePrompts,
} from './_shared.js';
import * as fs from 'fs';
import * as path from 'path';

// MARK: - Configuration (edit here)

const metaOverrides: Record<string, string> = {
  openSimModel:    'LaiUhlrich2022_shoulder', // 'LaiUhlrich2022_shoulder' | 'LaiUhlrich2022'
  scalingsetup:    'upright_standing_pose',   // 'upright_standing_pose' | 'any_pose'
  coreengine:      'v1.0',                    // 'v0.2' | 'v0.3' | 'v1.0'
  filterfrequency: 'default',                 // 'default' | integer Hz as string
};

// Per-trial overrides keyed by trial name.
// Setting 'activity_type' will auto-launch analysis after processing.
const trialMetaOverrides: Record<string, { activity_type?: string }> = {};
// Example:
// const trialMetaOverrides: Record<string, { activity_type?: string }> = {
//   trial1: { activity_type: ActivityType.RangeOfMotion },
//   trial2: { activity_type: ActivityType.CounterMovementJump },
// };

// MARK: - Settings mapping

function openSimModelFrom(value: string | undefined): SessionOpenSimModel {
  return value === 'LaiUhlrich2022' ? 'LaiUhlrich2022' : 'LaiUhlrich2022_shoulder';
}

function scalingSetupFrom(value: string | undefined): SessionScalingSetup {
  return value === 'any_pose' ? 'any_pose' : 'upright_standing_pose';
}

function coreEngineFrom(value: string | undefined): SessionCoreEngine {
  if (value === 'v0.2') return 'v0.2';
  if (value === 'v0.3') return 'v0.3';
  return 'v1.0';
}

function filterFrequencyFrom(value: string | undefined): FilterFrequency {
  if (!value || value === 'default') return { type: 'default' };
  const hz = parseInt(value, 10);
  return isNaN(hz) ? { type: 'default' } : { type: 'hz', value: hz };
}

// MARK: - OpenCap API

type OpenCapTrial = Record<string, unknown>;
type OpenCapSession = { trials: OpenCapTrial[]; subject: string; meta?: unknown };

async function fetchOpenCapSession(id: string, token: string, baseURL = 'https://api.opencap.ai/', retries = 3): Promise<OpenCapSession> {
  let lastErr: unknown;
  for (let attempt = 0; attempt < retries; attempt++) {
    try {
      const res = await fetch(`${baseURL}sessions/${id}/`, {
        headers: { Authorization: `Token ${token}` },
      });
      if (res.status === 500) throw new Error('No server response. Likely not a valid session ID.');
      const json = await res.json() as Record<string, unknown>;
      if (!json['trials']) throw new Error('Session not found or not accessible.');
      const trials = (json['trials'] as OpenCapTrial[]).slice().sort(
        (a, b) => String(a['created_at'] ?? '') < String(b['created_at'] ?? '') ? -1 : 1
      );
      return { ...json, trials } as OpenCapSession;
    } catch (err) {
      lastErr = err;
      if (attempt < retries - 1) {
        console.log(`Connection error fetching session, retrying in 5s: ${err}`);
        await sleep(5_000);
      }
    }
  }
  throw lastErr;
}

async function fetchOpenCapSubject(id: string, token: string, baseURL = 'https://api.opencap.ai/', retries = 3): Promise<Record<string, unknown>> {
  let lastErr: unknown;
  for (let attempt = 0; attempt < retries; attempt++) {
    try {
      const res = await fetch(`${baseURL}subjects/${id}/`, {
        headers: { Authorization: `Token ${token}` },
      });
      if (res.status === 500) throw new Error('No server response. Likely not a valid subject ID.');
      return await res.json() as Record<string, unknown>;
    } catch (err) {
      lastErr = err;
      if (attempt < retries - 1) {
        console.log(`Connection error fetching subject, retrying in 5s: ${err}`);
        await sleep(5_000);
      }
    }
  }
  throw lastErr;
}

function pickTrial(trials: OpenCapTrial[], predicate: (name: string) => boolean): OpenCapTrial | null {
  const matches = trials.filter(t => predicate(String(t['name'] ?? '').toLowerCase()));
  const done = matches.filter(t => t['status'] === 'done');
  const pool = done.length ? done : matches;
  if (!pool.length) return null;
  return pool.reduce((best, t) =>
    String(t['created_at'] ?? '') > String(best['created_at'] ?? '') ? t : best
  );
}

// MARK: - OpenCap login

async function opencapLogin(): Promise<string> {
  console.log('No OpenCap token found in .env or environment.');
  console.log('Log in with the credentials you use at app.opencap.ai.\n');
  const username = (await prompt('OpenCap username: ')).trim();
  const password = (await prompt('OpenCap password: ')).trim();

  const body = new URLSearchParams({ username, password });
  const res = await fetch('https://api.opencap.ai/login/', {
    method: 'POST',
    body,
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
  });
  const json = await res.json() as Record<string, unknown>;
  if (!json['token']) throw new Error('OpenCap login failed: unexpected response.');
  const token = String(json['token']);

  const envPath = path.join(process.cwd(), '.env');
  let envLines: string[] = [];
  if (fs.existsSync(envPath)) {
    envLines = fs.readFileSync(envPath, 'utf-8').split('\n');
  }
  envLines = envLines.filter(l => !l.startsWith('OPENCAP_TOKEN'));
  if (envLines.length && !envLines[envLines.length - 1].endsWith('\n')) envLines.push('');
  envLines.push(`OPENCAP_TOKEN="${token}"`);
  fs.writeFileSync(envPath, envLines.join('\n') + '\n');
  console.log(`Login successful. Token saved to ${envPath}.`);
  return token;
}

// MARK: - Copy session

async function copySession(
  service: ModelHealthService,
  sessionId: string,
  token: string,
  apiURL = 'https://api.opencap.ai/'
): Promise<Session> {
  console.log(`Fetching OpenCap session ${sessionId}...`);
  const sourceSession = await fetchOpenCapSession(sessionId, token, apiURL);

  const meta = (typeof sourceSession.meta === 'object' && sourceSession.meta !== null
    ? sourceSession.meta : {}) as Record<string, unknown>;
  const settings = Object.assign(
    {},
    (meta['settings'] as Record<string, string> | undefined) ?? {},
    metaOverrides
  );
  if (Object.keys(metaOverrides).length) console.log('Applying settings overrides.');

  const config: SessionConfig = {
    opensimModel:    openSimModelFrom(settings['openSimModel']),
    scalingSetup:    scalingSetupFrom(settings['scalingsetup']),
    coreEngine:      coreEngineFrom(settings['coreengine']),
    filterFrequency: filterFrequencyFrom(settings['filterfrequency']),
  };
  console.log('Session settings configured.');

  const subjects = await service.subjectList();

  let subject: Subject;
  if (subjects.length > 0 && await confirm(`Found ${subjects.length} subject(s). Select an existing one?`, true)) {
    console.log();
    subject = await pickOne(subjects, 'Select subject', s => `${s.name}  (ID ${s.id})`);
    console.log(`  Using: ${subject.name}`);
  } else {
    console.log('Fetching subject from OpenCap session...');
    const subjectData = await fetchOpenCapSubject(sourceSession.subject, token, apiURL);
    const name = String(subjectData['name'] ?? subjectData['first_name'] ?? '').trim() || 'Unknown';
    const weight = parseFloat(String(subjectData['weight'] ?? '0')) || 0;
    const heightM = parseFloat(String(subjectData['height'] ?? '0')) || 0;
    const birthYear = subjectData['birth_year'] ? Number(subjectData['birth_year']) : undefined;
    subject = await service.createSubject({ name, weight, height: heightM * 100, birthYear });
    console.log('Subject created successfully.');
  }

  const trials = sourceSession.trials;

  let calibrationTrial: OpenCapTrial | null = null;
  const sessionWithCal = (meta['sessionWithCalibration'] as Record<string, unknown> | undefined);
  const parentId = sessionWithCal?.['id'] as string | undefined;
  if (parentId) {
    try {
      const parentSession = await fetchOpenCapSession(parentId, token, apiURL);
      calibrationTrial = pickTrial(parentSession.trials, n => n === 'calibration');
      if (calibrationTrial) console.log(`Using calibration from parent session ${parentId}.`);
    } catch (err) {
      console.log(`Could not load calibration from parent session: ${err}`);
    }
  }
  if (!calibrationTrial) {
    calibrationTrial = pickTrial(trials, n => n === 'calibration');
  }

  if (!calibrationTrial) { console.error('No calibration trial found, cannot proceed.'); process.exit(1); }
  if (calibrationTrial['status'] !== 'done') { console.error('Calibration trial is not done, cannot proceed.'); process.exit(1); }

  const neutral = pickTrial(trials, n => n.includes('neutral'));
  if (!neutral) { console.error('No neutral trial found, cannot proceed.'); process.exit(1); }
  if (neutral['status'] !== 'done') { console.error('Neutral trial is not done, cannot proceed.'); process.exit(1); }

  const dynamicTrials = trials.filter(t => {
    const name = String(t['name'] ?? '').toLowerCase();
    return name !== 'calibration' && name !== 'neutral';
  });

  const annotated = dynamicTrials.map(trial => {
    const name = String(trial['name'] ?? '');
    const overrides = trialMetaOverrides[name];
    if (!overrides?.activity_type) return trial;
    return { ...trial, activity_type: overrides.activity_type };
  });

  const allTrials = [calibrationTrial, neutral, ...annotated];

  const seenUploads: Record<string, number> = {};
  let currentTrial: string | null = null;

  const onStatus = (status: ImportStatus) => {
    if (status.type === 'creating_session') {
      console.log('  Creating session...');
    } else if (status.type === 'created_session') {
      console.log(`  Session created: ${status.session_id}`);
    } else if (status.type === 'uploading_video' && status.uploaded < status.total) {
      currentTrial = status.trial;
      if (seenUploads[status.trial] !== status.uploaded) {
        seenUploads[status.trial] = status.uploaded;
        console.log(`  [${status.trial}] Uploading video ${status.uploaded + 1}/${status.total}...`);
      }
    } else if (status.type === 'processing') {
      const t = currentTrial ?? 'activity';
      console.log(`  [${t}] Processing...`);
    }
  };

  console.log('Importing session...');
  const importedSession = await service.importSession(
    JSON.stringify(allTrials),
    subject,
    config,
    onStatus,
  );
  console.log(`All activities imported successfully to session: ${importedSession.id}`);
  return importedSession;
}

// MARK: - Wait for activities

async function waitForActivities(service: ModelHealthService, session: Session): Promise<void> {
  let pending = session.activities.filter(a => {
    const name = (a.name ?? '').toLowerCase();
    return name !== 'calibration' && name !== 'neutral';
  });

  if (!pending.length) return;

  console.log('Waiting for activities to process...');
  const analyzing: Array<[Activity, Analysis]> = [];

  while (pending.length) {
    const stillPending: Activity[] = [];
    for (const activity of pending) {
      const label = activity.name ?? activity.id;
      let status: ActivityStatus;
      try {
        status = await service.activityStatus(activity);
      } catch (err) {
        console.log(`  [${label}] Status check failed: ${err}`);
        continue;
      }

      if (status.type === 'uploading') {
        console.log(`  [${label}] Uploading (${status.uploaded}/${status.total})...`);
        stillPending.push(activity);
      } else if (status.type === 'processing') {
        console.log(`  [${label}] Processing...`);
        stillPending.push(activity);
      } else if (status.type === 'analyzing') {
        console.log(`  [${label}] Analyzing — queued for analysis poll.`);
        analyzing.push([activity, { taskId: status.taskId }]);
      } else if (status.type === 'ready') {
        console.log(`  [${label}] Ready.`);
      } else {
        console.log(`  [${label}] Failed.`);
      }
    }
    if (stillPending.length) await sleep(5_000);
    pending = stillPending;
  }

  if (!analyzing.length) return;

  console.log('Waiting for analyses to complete...');
  let pendingAnalysis = analyzing;
  while (pendingAnalysis.length) {
    const stillPending: Array<[Activity, Analysis]> = [];
    for (const [activity, task] of pendingAnalysis) {
      const label = activity.name ?? activity.id;
      const status = await service.analysisStatus(task);
      if (status.type === 'processing') {
        console.log(`  [${label}] Analyzing...`);
        stillPending.push([activity, task]);
      } else if (status.type === 'completed') {
        console.log(`  [${label}] Analysis complete.`);
      } else {
        console.log(`  [${label}] Analysis failed.`);
      }
    }
    if (stillPending.length) await sleep(5_000);
    pendingAnalysis = stillPending;
  }
}

// MARK: - CLI helpers

async function resolveSessionId(args: string[]): Promise<string> {
  if (args[1] && args[1].trim()) return args[1].trim();
  const fromEnv = dotEnvValue('OPENCAP_SESSION_ID') ?? process.env.OPENCAP_SESSION_ID;
  if (fromEnv) return fromEnv;
  const id = (await prompt('OpenCap session ID: ')).trim();
  if (!id) { console.error('OpenCap session ID is required.'); process.exit(1); }
  return id;
}

async function resolveOpenCapToken(args: string[]): Promise<string> {
  if (args[2] && args[2].trim()) return args[2].trim();
  const fromEnv = dotEnvValue('OPENCAP_TOKEN') ?? process.env.OPENCAP_TOKEN;
  if (fromEnv) return fromEnv;
  return opencapLogin();
}

// MARK: - Main

async function main() {
  const args = process.argv.slice(2);

  const apiKey   = loadApiKey(args[0]);
  const sessionId = await resolveSessionId(args);
  const token     = await resolveOpenCapToken(args);

  console.log('Connecting to Model Health...');
  const service = new ModelHealthService({ apiKey, autoInit: false });
  await service.init();

  const session = await copySession(service, sessionId, token);
  await waitForActivities(service, session);
  console.log(`Import complete. Session ID: ${session.id}`);
}

main().catch(err => { console.error(`Error: ${err.message ?? err}`); process.exit(1); })
  .finally(closePrompts);
